### PR TITLE
First implementation

### PR DESCRIPTION
**Pull request by mattdean-digicatapult**
This PR merges the first functional version of the IPFS server we will use. The core of this is a nodejs application which will manage a runnning IPFS node setting and updating the swarm key to use as found in the `ipfsKey` module of the vitalam-node (see https://github.com/digicatapult/vitalam-node/pull/9 for details). A few points of note:

* Testing here is a bit hit and miss. I've implemented core testing of the restart functionality for now and a small amount of unit tests for the `keyWatcher` module that actually has some logic in it
* We assume that `go-ipfs` will be used. In dev mode we include the npm dev dependency `go-ipfs` which will pull a functional version of ipfs. For docker, because we are using alpine as a base, we build go-ipfs as part of the docker build
* The dockerfile assumes a persistant volume will be mounted at `/ipfs` and initialised separately. No tooling is provided to do this by the image and a separate go-ipfs image should likely be used.
* Helm chart is non-functional and incomplete. This will be done in another PR. For now relevant sections of the workflows surrounding the helm chart are commented out.

Note there are almost certainly issues here. Please review carefully and we'll discuss any issues in the approach here that will almost certainly come up

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+node_modules
+package.json
+coverage
+.nyc_output
+jsHelper.js
```
Comment (danielcooperxyz) [inline .eslintignore:5]: 👀

Comment (mattdean-digicatapult) [reply .eslintignore:5]: ah the daemon of copy pasta

File: package.json [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    "start": "node app/index.js",
+    "dev": "NODE_ENV=dev nodemon app/index.js | pino-colada",
+    "coverage": "nyc npm run test"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/digicatapult/vitalam-ipfs.git"
+  },
+  "engines": {
+    "node": ">=16.13.0"
```
Comment (darrylmorton-digicatapult) [inline package.json:21]: We should probably have another ticket in the backlog for updating the `vitalam-service-template` repo to node `v16.x`

**Review by darrylmorton-digicatapult — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (mattdean-digicatapult) [reply package.json:21]: 👍  I've already put a ticket in our Jira backlog for that IN-74

File: Dockerfile [hunk 3]
```diff
@@ -1,19 +1,53 @@
...
+ARG IPFS_TAG="v0.11.0"
+
+ENV SRC_DIR /go/src/github.com/ipfs/go-ipfs
+
+RUN apk add --no-cache git make bash gcc musl-dev
+  # && go get -u github.com/whyrusleeping/gx
+
+# Fixes an issue with symlinked binaries not playing well with musl
+# Thanks https://stackoverflow.com/questions/34729748/installed-go-binary-not-found-in-path-on-alpine-linux-docker
+# RUN mkdir /lib64 && ln -s /lib/libc.musl-x86_64.so.1 /lib64/ld-linux-x86-64.so.2
```
[the remark this answers is not in the document]
Comment (mattdean-digicatapult) [reply Dockerfile]: Yep cleaned up

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+coverage
+.DS_Store
+.nyc_output
+.idea
+.env
+.vscode
+
+github_token
+helm/**/*.tgz
+jsHelper.js
```
[the remark this answers is not in the document]
Comment (mattdean-digicatapult) [reply .gitignore]: Removed

File: README.md [hunk 5]
```diff
@@ -1,3 +1,21 @@
...
+## Environment Variables
+
+`vitalam-ipfs` is configured primarily using environment variables as follows:
+
+| variable                      | required | default | description                                                                                     |
+| :---------------------------- | :------: | :-----: | :---------------------------------------------------------------------------------------------- |
+| SERVICE_TYPE                  |    N     | `VITALAM_IPFS`  | Service type to appear in logs |
+| PORT                          |    N     | `80`  | The port for the API to listen on                                                               |
+| LOG_LEVEL                     |    N     | `info`  | Logging level. Valid values are [`trace`, `debug`, `info`, `warn`, `error`, `fatal`]            |
+| NODE_HOST | Y |  | Hostname of the `vitalam-node` to use as the swarm key source |
```
[the remark this answers is not in the document]
Comment (mattdean-digicatapult) [reply README.md]: Formatted

File: Dockerfile [hunk 6]
```diff
@@ -1,19 +1,48 @@
...
 
+# Expose 80 for healthcheck
 EXPOSE 80
-CMD ["node", "./app/index.js"]
+# Expose 4001 for ipfs swarm
+EXPOSE 4001
+# expose 5001 for ipfs api
+EXPOSE 5001
+
+HEALTHCHECK CMD curl --fail http://localhost:80/health || exit 1
```
Comment (dblane-digicatapult) [inline Dockerfile:46]: Necessary?

**Review by dblane-digicatapult — APPROVED** [review]

**Review by darrylmorton-digicatapult — APPROVED** [review]

**Pull request merged**